File: src/main/java/uk/co/rafearnold/randomworkoutroutine/service/GroupService.java
package uk.co.rafearnold.randomworkoutroutine.service;

import uk.co.rafearnold.randomworkoutroutine.model.entity.ExerciseOption;
import uk.co.rafearnold.randomworkoutroutine.model.entity.Group;
import uk.co.rafearnold.randomworkoutroutine.repository.GroupRepository;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Service
public class GroupService extends ItemService<Group> {

    private final RoutineService routineService;

    public GroupService(GroupRepository repository, RoutineService routineService) {
        super(repository);
        this.routineService = routineService;
    }

    public void delete(Group item) {
        routineService.removeGroupFromAll(item);
        super.delete(item);
    }

    public void removeExerciseFromAll(ExerciseOption exercise) {
        List<Group> groups = ((GroupRepository) repository).findAllByExerciseOptionsContaining(exercise);
        for (Group group : groups) {
            group.setExerciseOptions(group.getExerciseOptions().stream()
                    .filter(e -> e.getId() != exercise.getId())
                    .collect(Collectors.toList()));
            repository.save(group);
        }
    }

    @Override
    protected Group createItem(UUID id, String name) {
        Group group = new Group();
        group.setId(id);
        group.setName(name);
        return group;
    }
}

File: web/src/main/kotlin/uk/co/rafearnold/randomworkoutroutine/web/service/GroupService.kt
package uk.co.rafearnold.randomworkoutroutine.web.service

import org.springframework.stereotype.Service
import uk.co.rafearnold.randomworkoutroutine.web.model.entity.ExerciseOption
import uk.co.rafearnold.randomworkoutroutine.web.model.entity.Group
import uk.co.rafearnold.randomworkoutroutine.web.model.entity.Routine
import uk.co.rafearnold.randomworkoutroutine.web.repository.GroupRepository

/**
 * Service for retrieving and manipulating [Group] entities.
 */
@Service
class GroupService(
    repository: GroupRepository,
    private val routineService: RoutineService
) : ItemService<Group>(repository) {

    /**
     * Extension of [ItemService.delete] that also removes [item] from all [Routine]s that contain
     * it.
     */
    override fun delete(item: Group) {
        routineService.removeGroupFromAll(item)
        super.delete(item)
    }

    /**
     * Removes [exercise] from every [Group] whose [Group.exercises] property contains it.
     */
    fun removeExerciseFromAll(exercise: ExerciseOption) {
        val groups = (repository as GroupRepository).findAllByExercisesContaining(exercise)
        for (group in groups) {
            group.exercises = group.exercises.filter { it.id !== exercise.id }.toMutableList()
            repository.save(group)
        }
    }
}
